### PR TITLE
test: set up cypress component testing and test footer

### PR DESCRIPTION
**Pull request by marktnoonan**
## Description

This PR configures Cypress component testing and adds a component test for the footer, as well as testing the footer from the home page end-to-end test. 

It also adds a custom `cy.validateLink()` command based on the needs of this project and does a few other little things, full details are below. 

Note: it reuses the same test code for the component test and the end-to-end test to aid in refactoring - the test code does not depend on any details of the underlying UI library.

## Related Issue

Resolves: #122
Resolves: #142

## Motivation and Context

Chad mentioned to me that there were some open issues for Cypress tests targeting various components. In a Next.js application there are some components and ideas that are best to test from end-to-end, but many tests can be written at the component level just fine.

## Detailed notes on changes (in order of the diff)

- Add some Cypress things to `.gitignore`
- set up component testing config for Next + webpack
- run `validateFooter` in homepage e2e test
- define custom `validateLink` command
  - might be too much? but comes in handy
- add `component-index.html`, this is where Cypress mounts component tests
- add `component.ts` - support file for component testing with custom mount command to wrap components in `ChakraProvider` so they have what they need when rendering
- add `validateFooter` helper function for desktop/mobile footer testing
  - this is DOM-only, framework/library agnostic test code, works in e2e and component tests
- update scripts for Cypress in `package.json`, and update cypress
- add `footer.cy.tsx` - component test that tests the footer at different sizes
- add some `data-cy` attributes to the footer to make it easy to separate mobile and desktop
  - I think a better solution would be to only have one set of elements in the DOM at a time, currently the layout that is not being shown is hidden with CSS only, so we need to do this little workaround
- changes to `tsconfig.json` to make Cypress related files work correctly with TS
  - I haven't noticed any negative side-effects to this change but it would be worth checking out before merging in case some types clash somewhere or something.

## How Has This Been Tested?

Use the new npm scripts to confirm:

- `npm run cypress:open:e2e` should start you out at the Choose a Browser screen in e2e mode. Pick a browser and open e2e testing, then open `homepage.cy.ts` and note that it calls the new footer validation util to test the footer
- `npm run cypress:open:component` should start you out at the Choose a Browser screen in component testing mode. Pick a browser and open component testing, then open `footer.cy.tsx` and note that it also calls the new footer validation util to test the footer in desktop and mobile sizes
- `npm run cypress:run:e2e` should run the homepage e2e test in headless mode, not open the cypress UI
- `npm run cypress:run:component` should run the footer component test in headless mode, not open the cypress UI

## Screenshots (if appropriate):

Both testing types of Cypress are configured now:

Cypress testing types

Component test running for footer on mobile size:

mobile footer test

Homepage e2e test also testing footer:

homepage end-to-end to test

### CODE REVIEW

File: src/components/organisms/footer/footer.tsx [hunk 1]
```diff
@@ -217,11 +217,11 @@ const Footer = () => {
   });
 
   return (
-    <footer className="flex justify-center w-full bg-[#134D82]">y
```
Comment (marktnoonan) [inline src/components/organisms/footer/footer.tsx:220]: This lil `y` is kinda hanging out in prod being cool, I am a bit sad to remove it tbh.

Screenshot 2023-05-13 at 8 12 00 PM

**Comment by chadstewart** [comment]
Hey @marktnoonan, there seems to be an issue with the deployment of this branch. The error is:

```
./cypress/support/component.ts:20:19
--
Type error: Cannot find module '../../src/styles/globals.css' or its corresponding type declarations.
```

Any idea what could be causing this?

File: src/components/organisms/footer/footer.tsx (hunk 1 above)
Comment (chadstewart) [reply src/components/organisms/footer/footer.tsx:220]: LOL Yeah, I must have added this by accident working on some other fixes and didn't catch it.

Ugh, I hate when that happens.

**Comment by marktnoonan** [comment]
@chadstewart It looks like there is a typescript check that doesn't run by default locally, but runs pre-deployment, TS needed to be told it is OK to import a CSS file, when I run `tsc` after adding the declaration file it is ok.

On thing I didn't do in this PR is add component tests to the github actions for this project - wanted to make sure you were OK with the addition of the Cypress component tests first.

**Comment by chadstewart** [comment]
> @chadstewart It looks like there is a typescript check that doesn't run by default locally, but runs pre-deployment, TS needed to be told it is OK to import a CSS file, when I run `tsc` after adding the declaration file it is ok.
> 
> On thing I didn't do in this PR is add component tests to the github actions for this project - wanted to make sure you were OK with the addition of the Cypress component tests first.

Nah, it's cool man. Honestly, you know what you're doing much more than I do in terms of Cypress so I trust your judgement on what should be added and the set up needed.

Only thing I'd ask is could you elaborate a bit more on the why of some of the config? To be honest, this is primarily for me cause I want to learn Cypress and just want to know why I'd want to do some of this stuff in future projects.

Otherwise, pretty much consider yourself Chief Cypress Officer here. 😂

**Pull request merged**